opm/simulators/wells/WellFilterCake.cpp

```diff
@@ -37,106 +37,171 @@ namespace Opm {
 
 template<class Scalar>
 void WellFilterCake<Scalar>::
-updateFiltrationParticleVolume(const WellInterfaceGeneric<Scalar>& well,
-                               const double dt,
-                               const Scalar conc,
-                               const std::size_t water_index,
-                               WellState<Scalar>& well_state)
+updatePostStep(const WellInterfaceGeneric<Scalar>& well,
+               WellState<Scalar>& well_state,
+               const double dt,
+               const Scalar conc,
+               const std::size_t water_index,
+               DeferredLogger& deferred_logger)
 {
-    if (!well.isInjector()) {
+    if (! well.isInjector() || well.wellEcl().injectorType() != InjectorType::WATER)
         return;
-    }
-
-    if (filtration_particle_volume_.empty()) {
-        const auto& ws = well_state.well(well.indexOfWell());
-        filtration_particle_volume_.assign(ws.perf_data.size(), 0.); // initializing to be zero
-    }
 
-    const auto injectorType = well.wellEcl().injectorType();
-    if (injectorType != InjectorType::WATER) {
-        return;
+    assert (conc > 0.);
+    auto& ws = well_state.well(well.indexOfWell());
+    const auto nperf = ws.perf_data.size();
+    if (prev_skin_factor_.empty()) {
+        prev_skin_factor_.assign(nperf, 0.);
+        skin_factor_.assign(nperf, 0.);
+        thickness_.assign(nperf, 0.);
+        prev_thickness_.assign(nperf, 0.);
     }
 
-    auto& ws = well_state.well(well.indexOfWell());
+    prev_skin_factor_ = skin_factor_;
+    prev_thickness_ = thickness_;
     ws.filtrate_conc = conc;
 
-    if (conc == 0.) {
-        return;
-    }
+    updateMultipliers(well, well_state, dt, water_index, deferred_logger);
+}
 
-    const auto& connection_rates = ws.perf_data.phase_rates;
+template<class Scalar>
+void WellFilterCake<Scalar>::
+updatePreStep(const WellInterfaceGeneric<Scalar>& well, DeferredLogger& deferred_logger)
+{
+    // Apply cleaning and reset any filter cake cleaning multipliers (even if the well is producing at this time)
+    applyCleaning(well, deferred_logger);
```

Comment:
I see the point, inj_fc_multiplier_ must be updated here, but I think I would prefer to just recompute it from the updated skin factor to avoid saving the filtrate volume from the previous step (cleaning is not very frequent, so there should be little overhead).